python/hidet/distributed/partition/analyze.py

```diff
@@ -0,0 +1,153 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, List, Set, Tuple
+import tqdm
+
+from mip import Model, BINARY, xsum, minimize, OptimizationStatus
+
+from hidet.graph import FlowGraph, Operator, Tensor
+
+from .rule import op_shard_rule_search
+from .shard import OpShardSpec, TensorShardSpec, connect, node_comm_cost
+
+# I copied it from compiled_graph.py
+def get_graph_weights(graph: FlowGraph) -> List[Tensor]:
+    """
+    Get the weights of the graph. All constant tensors used by the operators in the graph, or returned directly by the
+    graph, are considered as weights.
+    """
+    weights: Set[Tensor] = set()
+    for node in graph.nodes:
+        for x in node.inputs:
+            if x.storage is not None:
+                weights.add(x)
+    for y in graph.outputs:
+        if y.storage is not None:
+            weights.add(y)
+    return list(weights)
+
+
+def generate_rules(g: FlowGraph, num_shards: int) -> Dict[Operator, List[OpShardSpec]]:
+    cache = {}
+    ret = {}
+    for node in tqdm.tqdm(g.nodes):
+        node_str = str(node)
+        if node_str not in cache:
+            cache[node_str] = op_shard_rule_search(node, num_shards)
```

Comment:
If node_str can uniquely identify the sharding rule, maybe cleaner to add `lru_cache` decorator  to `op_shard_rule_search` function?